### PR TITLE
Backward function for torch.cdist

### PR DESCRIPTION
**Pull request by ifedan**

### CODE REVIEW

File: aten/src/ATen/native/cuda/DistanceKernel.cu [hunk 1]
```diff
@@ -270,10 +301,50 @@ void pdist_backward_kernel_impl(Tensor& result, const Tensor& grad, const Tensor
...
+
 } // anonymous namespace
 
 REGISTER_DISPATCH(pdist_forward_stub, &pdist_forward_kernel_impl);
 REGISTER_DISPATCH(pdist_backward_stub, &pdist_backward_kernel_impl);
 REGISTER_DISPATCH(cdist_stub, &cdist_kernel_impl);
+REGISTER_DISPATCH(cdist_backward_stub, &cdist_backward_kernel_impl);
 
-}} // at::native
+}} // at::native
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/DistanceKernel.cu]: nit: I think there should be a newline here.

File: test/test_autograd.py [hunk 2]
```diff
@@ -2094,6 +2094,16 @@ def test_cat_empty_legacy(self):
                               False, f_args_variable, f_args_tensor)
         self.assertTrue(gradcheck(lambda a, b: torch.cat((a, b)), f_args_variable, eps=1e-6, atol=PRECISION))
 
+    def test_cdist(self):
```
Comment (gchanan) [inline test/test_autograd.py]: can you move this above or below all the `cat` tests?

File: test/test_autograd.py [hunk 3]
```diff
@@ -2094,6 +2094,16 @@ def test_cat_empty_legacy(self):
...
+        for p in [0, 1, 2, 3, 1.5, 2.5, float('inf')]:
+            f_args_variable = (torch.randn(S, S, requires_grad=True),
+                               torch.randn(S, S, requires_grad=True))
+            f = lambda a, b: torch.cdist(a, b, p)
+            f_args_tensor = deepcopy(unpack_variables(f_args_variable))
+            run_functional_checks(self, "test_cdist", "cdist", f,
+                                  True, f_args_variable, f_args_tensor)
```
Comment (gchanan) [inline test/test_autograd.py]: hmm...this exposed a pretty serious bug in our testing.  This test is running double backwards tests and passing, but it shouldn't be because you didn't write a double backwards function.

I believe what is happening is this:
1) because you don't have an entry in derivatives.yaml for your backwards function (i.e. a double backwards definition), it is assumed to be composed of differentiable functions
2) it's not actually composed of differentiable functions
3) the testing doesn't catch this because it only looks at differentiable outputs.  But your function doesn't create any differentiable outputs.

This should be testable, though.  Basically, this should only be valid if the numerical gradient of the function is 0.  Can you add checks for that?  (You should probably do it in a separate commit, there may be other functions that exhibit the same problem).

File: aten/src/ATen/core/aten_interned_strings.h [hunk 4]
```diff
@@ -103,6 +103,7 @@ _(aten, _pack_padded_sequence_backward) \
 _(aten, _pad_packed_sequence) \
 _(aten, _pdist_backward) \
 _(aten, _pdist_forward) \
+_(aten, _cdist_backward) \
```
[the remark this answers is not in the document]
Comment (ifedan) [reply aten/src/ATen/core/aten_interned_strings.h]: Removed

File: test/test_autograd.py [hunk 5]
```diff
@@ -2094,6 +2094,16 @@ def test_cat_empty_legacy(self):
...
+            self.assertTrue(gradcheck(f, f_args_variable, eps=1e-6, atol=PRECISION))
```
[the remark this answers is not in the document]
Comment (ifedan) [reply test/test_autograd.py]: Removed

File: aten/src/ATen/native/cuda/DistanceKernel.cu (hunk 1 above)
Comment (ifedan) [reply aten/src/ATen/native/cuda/DistanceKernel.cu]: Fixed

File: test/test_autograd.py (hunk 2 above)
Comment (ifedan) [reply test/test_autograd.py]: Fixed

File: test/test_autograd.py (hunk 3 above)
Comment (ifedan) [reply test/test_autograd.py]: Added test to EXCLUDE_GRADGRADCHECK_BY_TEST_NAME list

**Review by gchanan — COMMENTED** [review]
can you fix the underlying issue with gradgradcheck?  Which is basically that if none of the outputs require_grad, we don't actually check them.  We should instead check that their numerical gradients are 0.

**Review by gchanan — APPROVED** [review]

**Comment by ezyang** [comment]
This diff fails lint
